### DISCUSSION TITLE
Type Error after upgrade to version 5.4.1 / em.populate(myEntity, ['relation1', 'relation2'])

### DISCUSSION THREAD

**Discussion by woltob** [mikro-orm/mikro-orm#3479] [Q&A]
**Describe the bug**
After upgrading to version 5.4.1 I receive the following error for the following code, which previously (v 5.3.x) worked fine:

Code:
```
[myEntity] = await this.getEm().populate(myEntity, [
        'relation1',
        'relation2',
      ]);
```

Error:

```
Type 'string' is not assignable to type 'never'.ts(2322)

```

Populate wants an array of never[] as the second parameter for populate().

Is there a quick fix available?

**Versions**

| Dependency | Version |
| - | - |
| node | v16.16.0 |
| ts-node | v10.8.1 |
| typescript | 4.7.4 |
| mikro-orm | 5.4.1 |
| postgres |

**Comment by B4nan** [comment]
Cant help much without reproductions, but note this is a TS error, you can always get around those via casting.

Try upgrading TS to 4.8 to see if that helps, otherwise provide complete repro so I can observe.

**Reply by woltob** [reply]
I upgraded tsc to 4.8.3 and am running mikro orm 5.4.1.
Here is the issue I am seeing:

![image](https://user-images.githubusercontent.com/24568645/189551771-0f3e14cd-6248-43fd-a23b-e95d324a4264.png)

![image](https://user-images.githubusercontent.com/24568645/189551804-fe47262c-ba24-4a49-ade5-5428cbc42909.png)

![image](https://user-images.githubusercontent.com/24568645/189551887-8276545a-d12f-4fe8-ae0f-3cd8681cf812.png)



It is clearly visible that this is an issue of the type, as it appeared after upgrading Mikro ORM: 

EntityManager.d.ts

```typescript

* Loads specified relations in batch. This will execute one query for each relation, that will populate it on all of the specified entities.
     */
    populate<T extends object, P extends string = never>(entities: T | T[], populate: AutoPath<T, P>[] | boolean, options?: EntityLoaderOptions<T, P>): Promise<Loaded<T, P>[]>;

```

and this one from typings.d.ts 

```
export declare type AutoPath<O, P extends string, E extends string = never, D extends Prev[number] = 5> = [
    D
] extends [never] ? string : P extends any ? (P & `${string}.` extends never ? P : P & `${string}.`) extends infer Q ? Q extends `${infer A}.${infer B}` ? A extends StringKeys<O, E> ? `${A}.${AutoPath<Defined<GetStringKey<O, A, E>>, B, E, Prev[D]>}` : never : Q extends StringKeys<O, E> ? (Defined<GetStringKey<O, Q, E>> extends unknown ? Exclude<P, `${string}.`> : never) | (StringKeys<Defined<GetStringKey<O, Q, E>>, E> extends never ? never : `${Q}.`) : StringKeys<O, E> : never : never;
```

**Reply by B4nan** [reply]
You are assigning to a variable that already exists, right? As you just do `[ent] = ...` and not `const [ent] = ...`. The type of this variable is probably your problem.

**Reply by woltob** [reply]
Okay, amazing, this solved the issue! Is there any way I can just overwrite the previous variable to avoid having to reassign it? It used to work prior to 5.4.1 with version 5.3.x.

**Reply by B4nan** [reply]
Can you try adding `as const` assertion to your populate array? I think this might be actually caused by TS failing to infer the literal type. 

```ts
[myEntity] = await this.getEm().populate(myEntity, [
        'relation1',
        'relation2',
      ] as const);
```

**Reply by woltob** [reply]
`as const` does not work and produces:

```
Argument of type 'readonly ["organizationContact.contact", "organizationContact.userContacts", "invitingOrganization"]' is not assignable to parameter of type 'boolean | never[]'.
  The type 'readonly ["organizationContact.contact", "organizationContact.userContacts", "invitingOrganization"]' is 'readonly' and cannot be assigned to the mutable type 'never[]'.ts(2345)
```

**Reply by woltob** [reply]
When trying to overwrite the existing var, I see this:

```
Type 'Loaded<OrganizationInvitation, string>' is not assignable to type 'Loaded<OrganizationInvitation, "invitingOrganization" | "organizationContact.userContacts" | "organizationContact.contact"> | null'.
``` 

So it could be the return type of the populate function.

**Reply by B4nan** [reply]
I really think this is really just about your custom setup. There are no broken tests in the ORM, neither in any example apps I have. Yes, TS 4.8 was very breaking, we can't do much about that, only adapt.

I would need to have a complete repro to better understand this, there are so many weird signals, like `Loaded<OrganizationInvitation, string>` should never happen, unless you pass the `string` type there explicitly. That has no meaning, it needs to be literal string type or `never`. Maybe we could support your use case, but without having a complete repro at hand, there is not much we can do.

**Reply by woltob** [reply]
Okay, thank you. 
Perhaps the issue come from some service functions. 
If my Entity is called `OrganizationInvitation`, can the return type of a function that finds or creates the `OrganizationInvitation` still return a `Promise<OrganizationInvitation>` or does this have to be `Promise<Loaded<OrganizationInvitation>>` now?

**Reply by B4nan** [reply]
The goal of the `Loaded` type is to disallow assining an entity without populated relations. E.g. in your service you use `Loaded<User, 'org'>` in some method parameter, which means the method reuqires you to pass user entity with `org` property populated. You can't pass just `User` into that, it needs to be `Loaded<User, 'org'>` (or its extension = it can have more properties populated, but `org` will be mandatory). If your service just accepts naked `User`, you can pass it the `Loaded` instance as that is a super type, its compatible.

If you dont populate anything, it will end up as `Loaded<T, never>` which is the same as just `T`.

**Reply by woltob** [reply]
Okay, thanks for the clarification. I guess this is an interesting note for the documentation.
